neuralpp/symbolic/expression.py

```diff
@@ -1,9 +1,47 @@
 from __future__ import annotations  # to support forward reference for recursive type reference
-from typing import Any, List
+
+from typing import List, Any, Optional, Tuple, Type, NewType
 from abc import ABC, abstractmethod
 
 
+class FunctionType:
+    def __init__(self, argument_types: List[Type], return_type: Type):
+        self._argument_types = argument_types
+        self._return_type = return_type
+
+    @property
+    def argument_types(self):
+        return self._argument_types
+
+    @property
+    def return_type(self):
+        return self._return_type
+
+    @property
+    def arity(self):
+        return len(self._argument_types)
+
+    def return_type_after_application(self, number_of_arguments: int) -> FunctionType | Type:
```

Comment:
I think there's 2 major reasons why `FunctionType` should not subclass `Type`:

1. There seems to be some strict rule in Python for subclassing `Type` or `type`. In particular, we have no way to initialize a class like the following
```
class Bar(type):
    def __init__(self, foo1, foo2):
        pass
```
In particular, any subclass of `Type`/`type` must have three arguments: a name, a tuple and a dict (haven't looked into the reason, but if you call `Bar("a", "b", {})` you'll get error complaining the second argument is not a tuple. If you do not call with 3 arguments, you'll also get complains.
```
class Bar(Type):
    def __init__(self, name, tuple_, dict_):
        pass

bar = Bar("a", (), {})
```

So we'll need to refactor `FunctionType`'s `__init__` method, at lease.

2. The subclass of typing has the following properties (https://stackoverflow.com/questions/49142505/subclassing-type-vs-object-in-python3):
```
# Its instances are types
MyClass = Bar('MyClass', (), {})

isinstance(MyClass, type) # True

# And it is a class
x = MyClass()

isinstance(x, MyClass) # True
```

Which I believe `FunctionType` should not have. It is more like an annotation of type instead of this usage: when we initialize a `FunctionType` we are creating one particular instance of "function type annotation", not a `class`. I believe it is the same reason why `isinstance(Callable, Type) == False` and `isinstance(List, Type) == False`.

Reply:
tldr: `FunctionType` is not `Type`